### PR TITLE
Optimize hessian

### PR DESCRIPTION
**Pull request by sunqm**
* Optimize 4-center integral computation CUDA kernel for hessian
* Optimize 4-center integral CUDA kernel for Gradients
* CPHF solver on CPU to improve GPU memory utilization

### CODE REVIEW

File: gpu4pyscf/hessian/rks.py [hunk 1]
```diff
@@ -50,19 +50,38 @@ def partial_hess_elec(hessobj, mo_energy=None, mo_coeff=None, mo_occ=None,
...
+    if with_k and omega != 0:
+        j_factor = 0.
+        omega = -omega # Prefer computing the SR part
+        if alpha == 0: # LR=0, only SR exchange
+            k_factor = hyb
+        elif hyb == 0: # SR=0, only LR exchange
+            # full range exchange was computed in the previous step
+            k_factor = -alpha
+        else: # SR and LR exchange with different ratios
+            k_factor = hyb - alpha # =beta
```
Comment (wxj6000) [inline gpu4pyscf/hessian/rks.py:80]: k_factor is simply `hyb - alpha`. The comments still explain the different cases.

File: gpu4pyscf/hessian/rks.py [hunk 2]
```diff
@@ -50,19 +50,38 @@ def partial_hess_elec(hessobj, mo_energy=None, mo_coeff=None, mo_occ=None,
...
-        de2 -= hyb * ek
-    if abs(omega) > 1e-10 and abs(alpha-hyb) > 1e-10:
+        if omega == 0:
+            k_factor = hyb
+        elif alpha == 0: # LR=0, only SR exchange
+            pass
+        elif hyb == 0: # SR=0, only LR exchange
+            k_factor = alpha
+        else: # SR and LR exchange with different ratios
+            k_factor = alpha
```
Comment (wxj6000) [inline gpu4pyscf/hessian/rks.py:66]: `k_factor = hyb`
`if omega != 0:`
      `k_factor = alpha`
for simplicity?

Comment (sunqm) [reply gpu4pyscf/hessian/rks.py:66]: Similar condition code is implemented in various place for clarity

File: gpu4pyscf/hessian/rks.py (hunk 1 above)
Comment (sunqm) [reply gpu4pyscf/hessian/rks.py:80]: This part computes SR only.

**Review by wxj6000 — APPROVED** [review]

**Pull request merged**